clients/drcachesim/drpt2trace/pt2ir.cpp

```diff
@@ -0,0 +1,395 @@
+/* **********************************************************
+ * Copyright (c) 2022 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include <sys/stat.h>
+#include <stdlib.h>
+#include <iostream>
+#include <string.h>
+#include <inttypes.h>
+#include <errno.h>
+#include <fstream>
+
+#include "intel-pt.h"
+#include "libipt-sb.h"
+#include "pt2ir.h"
+extern "C" {
+#include "load_elf.h"
+}
+#define DR_FAST_IR 1
+#include "dr_api.h"
+
+#define ERRMSG_HEADER()                       \
+    do {                                      \
+        fprintf(stderr, "[drpt2ir] ERROR: "); \
+    } while (0)
+
+#define ERRMSG(...)                   \
+    do {                              \
+        ERRMSG_HEADER();              \
+        fprintf(stderr, __VA_ARGS__); \
+        fflush(stderr);               \
+    } while (0)
+
+pt2ir_t::pt2ir_t(IN const pt2ir_config_t config)
+    : config_(config)
+    , pt_raw_buffer_(nullptr)
+    , pt_raw_buffer_size_(0)
+    , pt_instr_decoder_(nullptr)
+    , pt_iscache_(nullptr)
+    , pt_sb_session_(nullptr)
+{
+    pt_iscache_ = pt_iscache_alloc(NULL);
```

Comment:
Use nullptr instead of NULL, here and everywhere in C++ code.

Reply:
Done.